### ISSUE TITLE
Improve/fix `IssueSchema.labels` / `MergeRequestSchema.labels` typing

### ISSUE THREAD

**Issue by kouak** [jdalrymple/gitbeaker#3384]
**Description**

`IssueSchema.labels` is typed as `undefined | string[]` which is correct unless `requestOptions.withLabelsDetails` is set to true.

In that case, Gitlab API does not return an array of string for this property, but an array of objects:

```typescript
const issues = await api.issues.all({ projectId: '...', withLabelsDetails: true });
```

```json5
[
  {
    id: 123,
    iid: 12345,
    // ...
    labels: [
      {
        id: 456,
        name: "bug",
        description: null,
        description_html: "",
        text_color: "#FFFFFF",
        color: "#000000",
      },
    ],
    // ...
  },
]
```

Merge requests have the same issue with the `withLabelDetails` request option.

**Proposal**

##### Easy fix: union type

Type `IssueSchema.labels` as an union : `string[] | IssueLabelDetailsSchema[]`.

This would generate typescript errors for current users.

```typescript
const issues = await api.Issues.all({ projectId: 'foo' });
for(const issue of issues) {
  for(const label of labels) {
    if(label.startsWith('foo')) { // Error !
      continue;
    }
    // ...
  }
}
```

##### Harder fix

IMO the ideal fix would be as follow:

```typescript
issues = await api.Issues.all({ projectId: 'foo' });
issues[0].labels; // undefined | string[]

issues = await api.Issues.all({ projectId: 'foo', withLabelDetails: true });
issues[0].labels; // undefined | IssueLabelDetailsSchema[];
```

I think I've got a solution with generics working for this kind of stuff (i.e, tweaking returned entity types based on request options). I'll follow this up with a draft PR for discussion.

**Checklist**
- [x] I have checked that this is not a duplicate issue.
- [x] I have read the documentation.

**Comment by jdalrymple** [comment]
Could also use a function overload? Might be significantly less work haha

**Comment by kouak** [comment]
I'm not too familiar with overloads in typescript, I usually stick with generics. Better the devil you know :)

I believe you need an overload per combination of input parameters, which could be a pain if there are multiple boolean flags ?

Anyway, here's a boiled down implementation with generics : https://tsplay.dev/N7VoDw

**Comment by jdalrymple** [comment]
The generics approach works, and is definitely powerful for similar situations, but it may be a bit overkill with only one variation for this type. If this type has multiple variations (beyond the labels), this would be the best way forward, however in this case there is just one. I also created a PR with an overload implementation. Give it a look and tell me if i missed something.

**Comment by jdalrymple** [comment]
:rocket: Issue was released in [`39.12.0`](https://github.com/jdalrymple/gitbeaker/releases/tag/39.12.0) :rocket:
